### DISCUSSION TITLE
Why CPD Doesn't work for Same file duplicates?

### DISCUSSION THREAD

**Discussion by thunderpycode** [pmd/pmd#5114] [Q&A] [answered]
We are running CPD, using "pmd cpd --minimum-tokens 10 --dir force-app_latest/main/default/classes --format xml --language apex",
We want to identify duplicate block of code within a same file but it is giving duplicates between the two files.
Can we use CPD to detect duplicate block of code within a same file ?
@himanshu22v

**Comment by thunderpycode** [comment]
@jsotuyod
Can you please answer this?

**Comment by jsotuyod** [comment]
CPD does report same file duplicates. Any duplicate code, irrespective of their location, will be reported.

Having said that, how the matching is done / what is considered to be a duplicate is subject to the parameters passed to CPD and what the language supports. In Apex that is currently limited to verbatim code copies, except for whitespace and comments that are ignored. That means variable names, literals, method names, etc. everything needs to match.

If you feel you have a scenario where this happens but is not detected, feel free to open an issue sharing the code snippets.

**Comment by thunderpycode** [comment]
Hi @jsotuyod,

we are trying to get report output in form of csv using below command in our pipeline script:

pmd cpd --minimum-tokens 100 --dir ${pmdFolder} --format csv --language apex --no-fail-on-violation 0 > cpd-report.csv",  returnStatus: true)

But couldn't get result as output file is not being created

@himanshu22v

**Reply by jsotuyod** [reply, marked as answer]
Output redirection is a shell feature. I don't know your pipeline tooling, but it may not be supporting it / executing the pmd binary directly without a shell.

Actually outputing to a file is something CPD should support, I'll file an issue for it (-> #4653). In the meantime, you may want to try actually running a bash in your pipeline, such as `bash -c "pmd cpd --minimum-tokens 100 --dir ${pmdFolder} --format csv --language apex --no-fail-on-violation 0 > cpd-report.csv"`

**Reply by thunderpycode** [reply]
Thanks @jsotuyod , we are using Jenkins declarative pipeline
